source/extensions/tracers/zipkin/zipkin_tracer_impl.cc

```diff
@@ -95,51 +96,22 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config, Http::HeaderMa
                                    const Tracing::Decision tracing_decision) {
   Tracer& tracer = *tls_->getTyped<TlsTracer>().tracer_;
   SpanPtr new_zipkin_span;
-  bool sampled(true);
-
-  if (request_headers.XB3Sampled()) {
-    // Checking if sampled flag has been specified. Also checking for 'true' value, as some old
-    // zipkin tracers may still use that value, although should be 0 or 1.
-    absl::string_view xb3_sampled = request_headers.XB3Sampled()->value().getStringView();
-    sampled = xb3_sampled == ZipkinCoreConstants::get().SAMPLED || xb3_sampled == "true";
-  } else {
-    sampled = tracing_decision.traced;
-  }
-
-  if (request_headers.XB3TraceId() && request_headers.XB3SpanId()) {
-    uint64_t trace_id(0);
-    uint64_t trace_id_high(0);
-    uint64_t span_id(0);
-    uint64_t parent_id(0);
-
-    // Extract trace id - which can either be 128 or 64 bit. For 128 bit,
-    // it needs to be divided into two 64 bit numbers (high and low).
-    if (request_headers.XB3TraceId()->value().size() == 32) {
-      const std::string tid = request_headers.XB3TraceId()->value().c_str();
-      const std::string high_tid = tid.substr(0, 16);
-      const std::string low_tid = tid.substr(16, 16);
-      if (!StringUtil::atoul(high_tid.c_str(), trace_id_high, 16) ||
-          !StringUtil::atoul(low_tid.c_str(), trace_id, 16)) {
-        return Tracing::SpanPtr(new Tracing::NullSpan());
-      }
-    } else if (!StringUtil::atoul(request_headers.XB3TraceId()->value().c_str(), trace_id, 16)) {
-      return Tracing::SpanPtr(new Tracing::NullSpan());
+  SpanContextExtractor extractor(request_headers);
+  bool sampled{extractor.extractSampled(tracing_decision)};
+  try {
+    auto ret_span_context = extractor.extractSpanContext(sampled);
+    if (!ret_span_context.second) {
+      // Create a root Zipkin span. No context was found in the headers.
```

Comment:
I didn't understand what you mean, extractSpanContext only in case of corrupt trace context will throw ExtractorException. Other case will return false

Reply:
Current code is fine to match existing behaviour - just suggesting that if a corrupt trace context is detected, we may want to cause a new trace instance (e.g. root span) to be created instead of a null span. But this could be addressed in a separate PR.